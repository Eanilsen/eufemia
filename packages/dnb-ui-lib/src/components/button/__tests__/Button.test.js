/**
 * Component Test
 *
 */

import React from 'react'
import {
  mount,
  fakeProps,
  toJson,
  axeComponent,
  loadScss
} from '../../../core/jest/jestSetup'
import Component from '../Button'
import dnb_button from '../style/dnb-button.scss' // eslint-disable-line
import dnb_button_theme_ui from '../style/themes/dnb-button-theme-ui.scss' // eslint-disable-line

const props = fakeProps(require.resolve('../Button'), {
  optional: true
})
props.variant = 'primary'
props.icon = 'question'
props.title = 'This is a button title'
props.size = null
props.icon_position = 'right'

describe('Button component', () => {
  it('have to match default button snapshot', () => {
    const Comp = mount(<Component {...props} href={null} />)
    expect(toJson(Comp)).toMatchSnapshot()
  })

  it('have to match href="..." snapshot', () => {
    const Comp = mount(<Component {...props} href="https://url" />)
    expect(toJson(Comp)).toMatchSnapshot()
  })

  it('has a button tag', () => {
    const title = 'title'
    const Comp = mount(<Component {...props} title={title} href={null} />)
    expect(Comp.find('button').exists()).toBe(true)
    expect(Comp.find('button').props().title).toBe(title)
  })

  it('icon only has to have some extra classes', () => {
    const Comp = mount(<Component icon="question" />)

    // size "medium" (32px)
    expect(Comp.find('.dnb-button--size-medium').exists()).toBe(true)

    // has icon class, but not has text
    expect(Comp.find('.dnb-button--has-icon').exists()).toBe(true)
    expect(Comp.find('.dnb-button--has-text').exists()).toBe(false)
  })

  it('has a button tag', () => {
    const Comp = mount(<Component {...props} href="https://url" />)
    expect(Comp.find('a').exists()).toBe(true)
  })

  it('should validate with ARIA rules as a button', async () => {
    const Comp = mount(<Component {...props} />)
    expect(await axeComponent(Comp)).toHaveNoViolations()
  })

  it('should validate with ARIA rules as a anchor', async () => {
    const Comp = mount(<Component {...props} href="https://url" />)
    expect(await axeComponent(Comp)).toHaveNoViolations()
  })
})

describe('Button scss', () => {
  it('have to match snapshot', () => {
    const scss = loadScss(require.resolve('../style/dnb-button.scss'))
    expect(scss).toMatchSnapshot()
  })
  it('have to match default theme snapshot', () => {
    const scss = loadScss(
      require.resolve('../style/themes/dnb-button-theme-ui.scss')
    )
    expect(scss).toMatchSnapshot()
  })
})
