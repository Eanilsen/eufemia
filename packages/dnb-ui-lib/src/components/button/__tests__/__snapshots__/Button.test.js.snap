// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component have to match default button snapshot 1`] = `
<Button
  attributes="attributes"
  class="class"
  className="className"
  custom_element={Object {}}
  custom_method={[Function]}
  disabled="disabled"
  href={null}
  icon="question"
  icon_position="right"
  icon_size="icon_size"
  id="id"
  innerRef={Object {}}
  on_click="on_click"
  size={null}
  text="text"
  title="This is a button title"
  type="type"
  variant="primary"
>
  <button
    className="dnb-button dnb-button--primary dnb-button--icon-position-right dnb-button--has-text dnb-button--has-icon id class className"
    id="id"
    onClick={[Function]}
    onMouseOut={[Function]}
    title="This is a button title"
    type="type"
  >
    <Content
      attributes="attributes"
      class="class"
      className="className"
      content="children"
      custom_element={Object {}}
      custom_method={[Function]}
      disabled="disabled"
      href={null}
      icon="question"
      icon_position="right"
      icon_size="icon_size"
      id="id"
      innerRef={Object {}}
      on_click="on_click"
      size={null}
      text="text"
      title="This is a button title"
      type="type"
      variant="primary"
    >
      children
      <span
        className="dnb-button__text"
        key="button-text"
      >
        text
      </span>
      <IconPrimary
        alt="This is a button title"
        area_hidden={true}
        attributes={null}
        class={null}
        className="dnb-button__icon"
        color={null}
        height={null}
        icon="question"
        key="button-icon"
        modifier={null}
        size="icon_size"
        width={null}
      >
        <span
          aria-hidden={true}
          aria-label="This is a button title"
          className="dnb-icon has-custom-size dnb-button__icon"
          role="img"
        >
          <question>
            <svg
              viewBox="0 0 16 16"
            >
              <path
                clipRule="evenodd"
                d="M4.352 5.1c0-2.05 1.55-3.6 3.6-3.6 2.052 0 3.601 1.55 3.601 3.6 0 2.051-1.55 3.6-3.6 3.6a.75.75 0 0 0-.75.75v2.666a.75.75 0 0 0 1.5 0V10.15c2.506-.348 4.35-2.425 4.35-5.049 0-2.879-2.22-5.1-5.1-5.1-2.88 0-5.1 2.221-5.1 5.1a.75.75 0 0 0 1.5 0zm3.6 10.898c.659 0 1.097-.439 1.097-1.097 0-.658-.438-1.096-1.096-1.096-.658 0-1.097.438-1.097 1.096 0 .658.439 1.097 1.097 1.097z"
                fillRule="evenodd"
              />
            </svg>
          </question>
        </span>
      </IconPrimary>
    </Content>
  </button>
</Button>
`;

exports[`Button component have to match href="..." snapshot 1`] = `
<Button
  attributes="attributes"
  class="class"
  className="className"
  custom_element={Object {}}
  custom_method={[Function]}
  disabled="disabled"
  href="https://url"
  icon="question"
  icon_position="right"
  icon_size="icon_size"
  id="id"
  innerRef={Object {}}
  on_click="on_click"
  size={null}
  text="text"
  title="This is a button title"
  type="type"
  variant="primary"
>
  <a
    className="dnb-button dnb-button--primary dnb-button--icon-position-right dnb-button--has-text dnb-button--has-icon id class className no-underline no-underline-hover"
    href="https://url"
    id="id"
    onClick={[Function]}
    onMouseOut={[Function]}
    title="This is a button title"
    type="type"
  >
    <Content
      attributes="attributes"
      class="class"
      className="className"
      content="children"
      custom_element={Object {}}
      custom_method={[Function]}
      disabled="disabled"
      href="https://url"
      icon="question"
      icon_position="right"
      icon_size="icon_size"
      id="id"
      innerRef={Object {}}
      on_click="on_click"
      size={null}
      text="text"
      title="This is a button title"
      type="type"
      variant="primary"
    >
      children
      <span
        className="dnb-button__text"
        key="button-text"
      >
        text
      </span>
      <IconPrimary
        alt="This is a button title"
        area_hidden={true}
        attributes={null}
        class={null}
        className="dnb-button__icon"
        color={null}
        height={null}
        icon="question"
        key="button-icon"
        modifier={null}
        size="icon_size"
        width={null}
      >
        <span
          aria-hidden={true}
          aria-label="This is a button title"
          className="dnb-icon has-custom-size dnb-button__icon"
          role="img"
        >
          <question>
            <svg
              viewBox="0 0 16 16"
            >
              <path
                clipRule="evenodd"
                d="M4.352 5.1c0-2.05 1.55-3.6 3.6-3.6 2.052 0 3.601 1.55 3.601 3.6 0 2.051-1.55 3.6-3.6 3.6a.75.75 0 0 0-.75.75v2.666a.75.75 0 0 0 1.5 0V10.15c2.506-.348 4.35-2.425 4.35-5.049 0-2.879-2.22-5.1-5.1-5.1-2.88 0-5.1 2.221-5.1 5.1a.75.75 0 0 0 1.5 0zm3.6 10.898c.659 0 1.097-.439 1.097-1.097 0-.658-.438-1.096-1.096-1.096-.658 0-1.097.438-1.097 1.096 0 .658.439 1.097 1.097 1.097z"
                fillRule="evenodd"
              />
            </svg>
          </question>
        </span>
      </IconPrimary>
    </Content>
  </a>
</Button>
`;

exports[`Button scss have to match snapshot 1`] = `
"/*
* DNB Button
*
* ======================================================================== */
/*
 * Utilities
 * ========================================================================== */
/*
 * Typography
 * ========================================================================== */
:root {
  --font-family-std: 'Fedra Sans Std', sans-serif;
  --font-family-demi: 'Fedra Sans Std Demi', sans-serif; }

.typo-std,
.typo-book {
  font-family: var(--font-family-std);
  font-weight: normal;
  font-style: normal; }

.typo-medium {
  font-family: var(--font-family-std);
  font-weight: 600;
  font-style: normal; }

.typo-demi {
  font-family: var(--font-family-demi);
  font-weight: normal;
  font-style: normal; }

.typo-number--old-style {
  font-variant-numeric: oldstyle-nums;
  -moz-font-feature-settings: \\"onum\\";
  -webkit-font-feature-settings: \\"onum\\";
  font-feature-settings: \\"onum\\"; }

.typo-number--lining {
  font-variant-numeric: lining-nums;
  -moz-font-feature-settings: \\"lnum\\";
  -webkit-font-feature-settings: \\"lnum\\";
  font-feature-settings: \\"lnum\\"; }

.typo-number--tabular {
  font-variant-numeric: tabular-nums;
  -moz-font-feature-settings: \\"tnum\\";
  -webkit-font-feature-settings: \\"tnum\\";
  font-feature-settings: \\"tnum\\"; }

.typo-number--proportional {
  font-variant-numeric: proportional-nums;
  -moz-font-feature-settings: \\"pnum\\";
  -webkit-font-feature-settings: \\"pnum\\";
  font-feature-settings: \\"pnum\\"; }

@font-face {
  font-family: 'Fedra Sans Std';
  src: url(\\"../../../../assets/fonts//FedraSansStd-Book.eot\\");
  src: url(\\"../../../../assets/fonts//FedraSansStd-Book.eot?#iefix\\") format(\\"embedded-opentype\\"), url(\\"../../../../assets/fonts//FedraSansStd-Book.woff2\\") format(\\"woff2\\"), url(\\"../../../../assets/fonts//FedraSansStd-Book.woff\\") format(\\"woff\\"), url(\\"../../../../assets/fonts//FedraSansStd-Book.ttf\\") format(\\"truetype\\");
  font-weight: normal;
  font-style: normal; }

@font-face {
  font-family: 'Fedra Sans Std';
  src: url(\\"../../../../assets/fonts//FedraSansStd-Medium.eot\\");
  src: url(\\"../../../../assets/fonts//FedraSansStd-Medium.eot?#iefix\\") format(\\"embedded-opentype\\"), url(\\"../../../../assets/fonts//FedraSansStd-Medium.woff2\\") format(\\"woff2\\"), url(\\"../../../../assets/fonts//FedraSansStd-Medium.woff\\") format(\\"woff\\"), url(\\"../../../../assets/fonts//FedraSansStd-Medium.ttf\\") format(\\"truetype\\");
  font-weight: 600;
  font-style: normal; }

@font-face {
  font-family: 'Fedra Sans Std Demi';
  src: url(\\"../../../../assets/fonts//FedraSansStd-Demi.eot\\");
  src: url(\\"../../../../assets/fonts//FedraSansStd-Demi.eot?#iefix\\") format(\\"embedded-opentype\\"), url(\\"../../../../assets/fonts//FedraSansStd-Demi.woff2\\") format(\\"woff2\\"), url(\\"../../../../assets/fonts//FedraSansStd-Demi.woff\\") format(\\"woff\\"), url(\\"../../../../assets/fonts//FedraSansStd-Demi.ttf\\") format(\\"truetype\\");
  font-weight: normal;
  font-style: normal; }

/*
* Scopes
* ======================================================================== */
/*
* DNB icon
*
* ======================================================================== */
/*
* Icon component
*
* ======================================================================== */
.dnb-icon {
  display: inline;
  vertical-align: baseline;
  font-size: inherit;
  width: 1em;
  height: 1em; }
  .dnb-icon svg {
    fill: currentColor; }
  .dnb-icon:not(.has-custom-size) svg {
    width: inherit;
    height: inherit; }
  .dnb-icon--small {
    width: 0.5em;
    height: 0.5em; }
  .dnb-icon--default {
    width: 1em;
    height: 1em; }
  .dnb-icon--medium {
    width: 1.5em;
    height: 1.5em; }
  .dnb-icon--large {
    width: 2em;
    height: 2em; }

.dnb-button {
  font-family: var(--font-family-std, sans-serif);
  font-size: 1rem;
  font-weight: normal;
  font-style: normal;
  line-height: 1.5rem;
  color: var(--color-black-80, #333); }
  .dnb-button *,
  .dnb-button ::before,
  .dnb-button ::after {
    background-repeat: no-repeat;
    box-sizing: border-box; }
  .dnb-button ::before,
  .dnb-button ::after {
    text-decoration: inherit;
    vertical-align: inherit; }

/*
* Button component
*
* ======================================================================== */
:root {
  --button-font-family: var(--font-family-std);
  --button-font-size: 1rem;
  --button-font-weight: normal;
  --button-font-style: normal;
  --button-width: 2.5rem;
  --button-height: 2.5rem;
  --button-width--medium: 2rem;
  --button-height--medium: 2rem;
  --button-width--large: 3rem;
  --button-height--large: 3rem;
  --button-icon-size: 1rem;
  --button-border-width: 1px;
  --button-border-width--hover: 3px;
  --button-border-radius: calc(var(--button-height) / 2);
  --button-border-radius--medium: calc(var(--button-height--medium) / 2);
  --button-border-radius--large: calc(var(--button-height--large) / 2); }

.dnb-button {
  display: inline-flex;
  vertical-align: middle;
  align-content: center;
  align-items: center;
  justify-content: center;
  white-space: nowrap;
  cursor: pointer;
  width: var(--button-width);
  height: var(--button-height);
  padding: 0;
  border: var(--button-border-width) solid transparent;
  border-radius: var(--button-border-radius);
  font-family: var(--button-font-family);
  font-size: var(--button-font-size);
  font-weight: var(--button-font-weight);
  font-style: var(--button-font-style);
  text-decoration: none;
  /* Firefox includes a hidden border which messes up button dimensions */ }
  .dnb-button__icon {
    line-height: 0; }
    .dnb-button__icon:not(.dnb-icon) svg:not([width]):not([height]) {
      width: var(--button-icon-size);
      height: var(--button-icon-size); }
  .dnb-button__text {
    color: inherit;
    background-color: inherit;
    transform: translateY(var(--button-border-width)); }
  .dnb-button--size-medium {
    width: var(--button-width--medium);
    height: var(--button-height--medium);
    border-radius: var(--button-border-radius--medium); }
  .dnb-button--has-text.dnb-button--size-medium {
    padding-left: 1em;
    padding-right: 1em; }
  .dnb-button--has-text.dnb-button--icon-position-left.dnb-button--size-medium {
    padding-left: 0.5em; }
  .dnb-button--has-text.dnb-button--icon-position-right.dnb-button--size-medium {
    padding-right: 0.5em; }
  .dnb-button--size-large {
    width: var(--button-width--large);
    height: var(--button-height--large);
    border-radius: var(--button-border-radius--large); }
  .dnb-button--has-text.dnb-button--size-large {
    padding-left: 2em;
    padding-right: 2em; }
  .dnb-button--has-text.dnb-button--icon-position-left.dnb-button--size-large {
    padding-left: 1em; }
  .dnb-button--has-text.dnb-button--icon-position-right.dnb-button--size-large {
    padding-right: 1em; }
  .dnb-button--has-text {
    padding-left: 1.5em;
    padding-right: 1.5em;
    width: auto; }
  .dnb-button--has-text .dnb-button__icon {
    margin: 0 calc(var(--button-icon-size) / 2); }
  .dnb-button--has-text.dnb-button--icon-position-left {
    padding-left: 0.5em; }
  .dnb-button--has-text.dnb-button--icon-position-right {
    padding-right: 0.5em; }
  .dnb-button--has-text.dnb-button--has-icon .dnb-button__icon {
    order: 2; }
  .dnb-button--has-text.dnb-button--has-icon .dnb-button__text {
    order: 1; }
  .dnb-button--has-text.dnb-button--has-icon.dnb-button--icon-position-left .dnb-button__icon {
    order: 1; }
  .dnb-button--has-text.dnb-button--has-icon.dnb-button--icon-position-left .dnb-button__text {
    order: 2; }
  .dnb-button--reset {
    margin: 0;
    padding: 0;
    width: auto;
    height: auto;
    overflow: visible;
    border: none;
    border-radius: 0;
    background-color: transparent;
    appearance: none;
    box-shadow: none;
    color: inherit;
    font: inherit;
    text-align: inherit;
    line-height: 0; }
    .no-touch .dnb-button--reset:hover[disabled] {
      cursor: not-allowed; }
    .no-touch .dnb-button--reset:hover:not([disabled]) {
      opacity: 0.6;
      box-shadow: none;
      border: none; }
    .dnb-button--reset:not([disabled]):focus, .dnb-button--reset:not([disabled]):active {
      outline: none;
      border-color: transparent; }
      html[data-whatinput='mouse'] .dnb-button--reset:not([disabled]):focus:not(.dnb-no-mouse-focus), html[data-whatinput='mouse'] .dnb-button--reset:not([disabled]):active:not(.dnb-no-mouse-focus) {
        transition: box-shadow 0.1s ease-out 0s;
        box-shadow: 0 0 0 2px var(--color-emerald-green); }
      html[data-whatinput='keyboard'] .dnb-button--reset:not([disabled]):focus, html[data-whatinput='keyboard'] .dnb-button--reset:not([disabled]):active {
        box-shadow: 0 0 0 2px var(--color-signal-orange); }
    html[data-whatinput='mouse']
.dnb-button--reset:not([disabled]):focus:not(.dnb-no-mouse-focus),
    html[data-whatinput='mouse']
.dnb-button--reset:not([disabled]):active:not(.dnb-no-mouse-focus) {
      box-shadow: none;
      color: inherit;
      border: none; }
  .dnb-button[type='button'] {
    appearance: none; }
  a .dnb-button {
    border-bottom: none; }
  button.dnb-button::-moz-focus-inner .dnb-button {
    border: none; }
"
`;
